File: frontend/lib/conversation.ts
/**
 * 会话管理 API
 */

import { API_BASE_URL } from "./api";

// ==================== 通用请求函数 ====================

async function request<T>(
    endpoint: string,
    options: RequestInit = {}
): Promise<T> {
    const url = `${API_BASE_URL}${endpoint}`;

    const response = await fetch(url, {
        ...options,
        headers: {
            "Content-Type": "application/json",
            ...options.headers,
        },
    });

    if (!response.ok) {
        const error = await response.text();
        throw new Error(error || `请求失败: ${response.status}`);
    }

    return response.json();
}

// ==================== 类型定义 ====================

export interface ChatSession {
    chat_session_id: string;
    title: string | null;
    folder_ids: number[] | null;
    message_count: number;
    last_message_at: string | null;
    created_at: string;
    is_archived: boolean;
}

export interface ChatMessage {
    id: number;
    role: "user" | "assistant";
    content: string;
    sources: Array<{
        bvid: string;
        title: string;
        url: string;
    }> | null;
    route: string | null;
    created_at: string;
}

export interface SessionListResponse {
    sessions: ChatSession[];
    total: number;
    page: number;
    page_size: number;
}

export interface MessageListResponse {
    messages: ChatMessage[];
    total: number;
    page: number;
    page_size: number;
}

export interface SessionCreateResponse {
    chat_session_id: string;
    title: string;
    created_at: string;
}

export interface SearchResult {
    chat_session_id: string;
    session_title: string | null;
    message_id: number;
    content: string;
    highlight: string | null;
    created_at: string;
}

export interface SearchResponse {
    results: SearchResult[];
    total: number;
    page: number;
    page_size: number;
}

// ==================== API 函数 ====================

export const conversationApi = {
    // 获取会话列表
    list: (
        userSessionId: string,
        page = 1,
        pageSize = 20,
        includeArchived = false
    ): Promise<SessionListResponse> =>
        request(
            `/conversation/list?user_session_id=${userSessionId}&page=${page}&page_size=${pageSize}&include_archived=${includeArchived}`
        ),

    // 创建新会话
    create: (
        userSessionId: string,
        title?: string,
        folderIds?: number[]
    ): Promise<SessionCreateResponse> =>
        request("/conversation/create", {
            method: "POST",
            body: JSON.stringify({
                user_session_id: userSessionId,
                title,
                folder_ids: folderIds,
            }),
        }),

    // 获取会话详情
    get: (chatSessionId: string, userSessionId: string): Promise<ChatSession> =>
        request(
            `/conversation/${chatSessionId}?user_session_id=${userSessionId}`
        ),

    // 更新会话
    update: (
        chatSessionId: string,
        userSessionId: string,
        data: { title?: string; is_archived?: boolean }
    ): Promise<ChatSession> =>
        request(`/conversation/${chatSessionId}?user_session_id=${userSessionId}`, {
            method: "PUT",
            body: JSON.stringify(data),
        }),

    // 删除会话
    delete: (chatSessionId: string, userSessionId: string): Promise<{ success: boolean; message: string }> =>
        request(`/conversation/${chatSessionId}?user_session_id=${userSessionId}`, {
            method: "DELETE",
        }),

    // 获取会话消息
    getMessages: (
        chatSessionId: string,
        userSessionId: string,
        page = 1,
        pageSize = 50
    ): Promise<MessageListResponse> =>
        request(
            `/conversation/${chatSessionId}/messages?user_session_id=${userSessionId}&page=${page}&page_size=${pageSize}`
        ),

    // 搜索对话
    search: (
        userSessionId: string,
        query: string,
        page = 1,
        pageSize = 20
    ): Promise<SearchResponse> =>
        request(
            `/conversation/search?user_session_id=${userSessionId}&query=${encodeURIComponent(query)}&page=${page}&page_size=${pageSize}`
        ),
};

File: app/routers/conversation.py
"""
Bilibili RAG 知识库系统
对话会话管理路由
"""
from typing import Optional
from fastapi import APIRouter, HTTPException, Depends, Query, Body
from loguru import logger
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models import (
    ChatSessionListResponse,
    ChatSessionInfo,
    ChatSessionCreateResponse,
    ChatSessionUpdateRequest,
    ChatMessageListResponse,
    ChatMessageInfo,
    ConversationSearchResponse,
    ConversationSearchResult,
)
from app.services.conversation import ConversationService


class ChatSessionCreateRequest(BaseModel):
    """创建会话请求"""
    user_session_id: str
    title: Optional[str] = None
    folder_ids: Optional[list[int]] = None


router = APIRouter(prefix="/conversation", tags=["会话管理"])


def get_conversation_service(db: AsyncSession = Depends(get_db)) -> ConversationService:
    """获取会话服务实例"""
    return ConversationService(db)


@router.get("/list", response_model=ChatSessionListResponse)
async def list_conversations(
    user_session_id: str = Query(..., description="用户登录 session"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    include_archived: bool = Query(False, description="是否包含已归档的会话"),
    service: ConversationService = Depends(get_conversation_service)
):
    """获取会话列表"""
    sessions, total = await service.list_sessions(
        user_session_id=user_session_id,
        page=page,
        page_size=page_size,
        include_archived=include_archived
    )

    session_infos = [
        ChatSessionInfo(
            chat_session_id=s.session_id,
            title=s.title,
            folder_ids=s.folder_ids,
            message_count=s.message_count,
            last_message_at=s.last_message_at,
            created_at=s.created_at,
            is_archived=s.is_archived
        )
        for s in sessions
    ]

    return ChatSessionListResponse(
        sessions=session_infos,
        total=total,
        page=page,
        page_size=page_size
    )


@router.post("/create", response_model=ChatSessionCreateResponse)
async def create_conversation(
    request: ChatSessionCreateRequest,
    service: ConversationService = Depends(get_conversation_service)
):
    """创建新会话"""
    session = await service.create_session(
        user_session_id=request.user_session_id,
        title=request.title,
        folder_ids=request.folder_ids
    )

    return ChatSessionCreateResponse(
        chat_session_id=session.session_id,
        title=session.title or "新会话",
        created_at=session.created_at
    )


@router.get("/{chat_session_id}", response_model=ChatSessionInfo)
async def get_conversation(
    chat_session_id: str,
    user_session_id: str = Query(..., description="用户登录 session"),
    service: ConversationService = Depends(get_conversation_service)
):
    """获取会话详情"""
    session = await service.get_session_by_user(chat_session_id, user_session_id)
    if not session:
        raise HTTPException(status_code=404, detail="会话不存在")

    return ChatSessionInfo(
        chat_session_id=session.session_id,
        title=session.title,
        folder_ids=session.folder_ids,
        message_count=session.message_count,
        last_message_at=session.last_message_at,
        created_at=session.created_at,
        is_archived=session.is_archived
    )


@router.put("/{chat_session_id}", response_model=ChatSessionInfo)
async def update_conversation(
    chat_session_id: str,
    user_session_id: str = Query(..., description="用户登录 session"),
    request: ChatSessionUpdateRequest = None,
    service: ConversationService = Depends(get_conversation_service)
):
    """更新会话（重命名、归档）"""
    session = await service.update_session(
        chat_session_id=chat_session_id,
        user_session_id=user_session_id,
        title=request.title if request else None,
        is_archived=request.is_archived if request else None
    )
    if not session:
        raise HTTPException(status_code=404, detail="会话不存在")

    return ChatSessionInfo(
        chat_session_id=session.session_id,
        title=session.title,
        folder_ids=session.folder_ids,
        message_count=session.message_count,
        last_message_at=session.last_message_at,
        created_at=session.created_at,
        is_archived=session.is_archived
    )


@router.delete("/{chat_session_id}")
async def delete_conversation(
    chat_session_id: str,
    user_session_id: str = Query(..., description="用户登录 session"),
    service: ConversationService = Depends(get_conversation_service)
):
    """删除会话（软删除）"""
    success = await service.delete_session(chat_session_id, user_session_id)
    if not success:
        raise HTTPException(status_code=404, detail="会话不存在")

    return {"success": True, "message": "会话已删除"}


@router.get("/{chat_session_id}/messages", response_model=ChatMessageListResponse)
async def get_conversation_messages(
    chat_session_id: str,
    user_session_id: str = Query(..., description="用户登录 session"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(50, ge=1, le=100, description="每页数量"),
    service: ConversationService = Depends(get_conversation_service)
):
    """获取会话消息列表"""
    # 验证会话归属
    session = await service.get_session_by_user(chat_session_id, user_session_id)
    if not session:
        raise HTTPException(status_code=404, detail="会话不存在")

    messages, total = await service.get_messages(
        chat_session_id=chat_session_id,
        page=page,
        page_size=page_size
    )

    message_infos = [
        ChatMessageInfo(
            id=m.id,
            role=m.role,
            content=m.content,
            sources=m.sources,
            route=m.route,
            created_at=m.created_at
        )
        for m in messages
    ]

    return ChatMessageListResponse(
        messages=message_infos,
        total=total,
        page=page,
        page_size=page_size
    )


@router.get("/search", response_model=ConversationSearchResponse)
async def search_conversations(
    user_session_id: str = Query(..., description="用户登录 session"),
    query: str = Query(..., min_length=1, description="搜索关键词"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    service: ConversationService = Depends(get_conversation_service)
):
    """搜索对话内容"""
    results, total = await service.search_messages(
        user_session_id=user_session_id,
        query=query,
        page=page,
        page_size=page_size
    )

    search_results = [
        ConversationSearchResult(
            chat_session_id=r["chat_session_id"],
            session_title=r["session_title"],
            message_id=r["message_id"],
            content=r["content"],
            highlight=r["highlight"],
            created_at=r["created_at"]
        )
        for r in results
    ]

    return ConversationSearchResponse(
        results=search_results,
        total=total,
        page=page,
        page_size=page_size
    )
